### PR TITLE
[Core] Adding Solve to MathUtils

### PR DESCRIPTION
**Pull request by loumalouomega**
The title says everything

### CODE REVIEW

File: kratos/utilities/math_utils.h [hunk 1]
```diff
@@ -67,14 +67,14 @@ class MathUtils
...
+
     typedef Matrix MatrixType;
 
     typedef Vector VectorType;
 
     typedef std::size_t SizeType;
-    
-    typedef unsigned int IndexType;
+
+    typedef std::size_t IndexType;
```
Comment (roigcarlo) [inline kratos/utilities/math_utils.h:77]: If we are gonna have two typedef with the same meaning I propose to delete one.

File: kratos/utilities/math_utils.h [hunk 2]
```diff
@@ -403,7 +403,31 @@ class MathUtils
...
-    
+
+    /**
+     * @brief This function is designed to be called when a dense linear system is needed to be solved
+     * @param rA System matrix
+     * @param rX Solution vector. it's also the initial guess for iterative linear solvers.
+     * @param rB Right hand side vector.
+     */
+    static void Solve(
+        const MatrixType& rA,
```
Comment (RiccardoRossi) [inline kratos/utilities/math_utils.h]: it is more elegant if instead of making the copy internally you pass A by value

File: kratos/utilities/math_utils.h [hunk 3]
```diff
@@ -403,7 +403,31 @@ class MathUtils
...
+        const SizeType size1 = rA.size1();
+
+        rX = rB;
+        typedef permutation_matrix<SizeType> pmatrix;
+        Matrix A(rA);
+        pmatrix pm(size1);
+        int singular = lu_factorize(A,pm);
+        lu_substitute(A, pm, rX);
+
+        KRATOS_ERROR_IF(singular == 1) << "::ERROR: Matrix is singular: " << rA << std::endl;
```
Comment (RiccardoRossi) [inline kratos/utilities/math_utils.h]: i would put this as KRATOS_DEBUG_ERROR_IF

in any case even if you don't put the error before the lu_substitute

File: kratos/utilities/math_utils.h [hunk 4]
```diff
@@ -441,32 +465,32 @@ class MathUtils
...
+
             // Calculating determinant
             InputMatrixDet = 1.0;
-            
-            for (unsigned int i = 0; i < A.size1();i++)
+
+            for (IndexType i = 0; i < A.size1();i++)
             {
-                unsigned int ki = pm[i] == i ? 0 : 1;
+                IndexType ki = pm[i] == i ? 0 : 1;
```
Comment (RiccardoRossi) [inline kratos/utilities/math_utils.h]: it is not really part of this PR, however since i see it i post it here

what is done here on the next line? this looks to me VERY inefficient ... the determinant should be directly the product of the diagonal terms, without the pow. are u really sure the pow is needed?

File: kratos/utilities/math_utils.h (hunk 1 above)
Comment (loumalouomega) [reply kratos/utilities/math_utils.h:77]: This is done in all Kratos, do a grep with IndexType and SizeType

Comment (roigcarlo) [reply kratos/utilities/math_utils.h:77]: ¯\\_(ツ)_/¯ Ok then but still think is missleading

**Review by roigcarlo — DISMISSED** [review]

**Review by RiccardoRossi — CHANGES_REQUESTED** [review]

File: kratos/utilities/math_utils.h (hunk 4 above)
Comment (loumalouomega) [reply kratos/utilities/math_utils.h]: I cannot remember, I did that long time ago :P

File: kratos/utilities/math_utils.h (hunk 3 above)
Comment (loumalouomega) [reply kratos/utilities/math_utils.h]: OK

File: kratos/utilities/math_utils.h (hunk 2 above)
Comment (loumalouomega) [reply kratos/utilities/math_utils.h]: OK

Comment (loumalouomega) [reply kratos/utilities/math_utils.h]: The lu_substitute modifies internally A, so better to do a copy, imagine that you are interested in keep the A matrix as it is

File: kratos/utilities/math_utils.h (hunk 4 above)
Comment (loumalouomega) [reply kratos/utilities/math_utils.h]: https://en.wikipedia.org/wiki/LU_decomposition#Computing_the_determinant according to this it is jot necessary

Comment (loumalouomega) [reply kratos/utilities/math_utils.h]:
~~~
    double det = A[0][0];

    for (int i = 1; i < N; i++)
        det *= A[i][i];

    if ((P[N] - N) % 2 == 0)
        return det; 
    else
        return -det;
~~~

Comment (RiccardoRossi) [reply kratos/utilities/math_utils.h]: in the code snyppet you put, the if is there only once... in the one you implemented it is there for every row...

File: kratos/utilities/math_utils.h (hunk 2 above)
Comment (RiccardoRossi) [reply kratos/utilities/math_utils.h]: correct, but you can pass it by value to the function... it is equivalent to your code but more elegant. anyway i will accept it as it is since it is not wrong

File: kratos/utilities/math_utils.h (hunk 4 above)
Comment (loumalouomega) [reply kratos/utilities/math_utils.h]: Yeah, but the P and N are unkown if I use the lu_substitute

Comment (loumalouomega) [reply kratos/utilities/math_utils.h]: That's why

Comment (loumalouomega) [reply kratos/utilities/math_utils.h]: In any case I got rid of the std::pow

**Review by RiccardoRossi — DISMISSED** [review]
accepting

**Comment by loumalouomega** [comment]
Ok, sorry, you mean without reference, one sec

**Review by roigcarlo — APPROVED** [review]
